src/Device/Driver/XCVario.cpp

```diff
@@ -138,43 +167,98 @@ PXCV(NMEAInputLine &line, NMEAInfo &info)
     info.attitude.pitch_angle = Angle::Degrees(value);
   }
   // Space diagonal acceleration in X,Y,Z axes measure
-  if ( line.ReadChecked(x) && line.ReadChecked(y) && line.ReadChecked(z) )
+  if (line.ReadChecked(x) && line.ReadChecked(y) && line.ReadChecked(z))
     info.acceleration.ProvideGLoad(SpaceDiagonal(x, y, z));
 
   return true;
 }
 
+bool
+XVCDevice::XCV(NMEAInputLine &line, NMEAInfo &info)
+{
+  double value;
+  char topic[16];
+  line.Read(topic, sizeof(topic));
+  if(StringIsEqual(topic, "bal-water")){
+    if (line.ReadChecked(value)){
+      info.settings.ProvideBallastLitres(value, info.clock);
+    }
+  }
+  else if(StringIsEqual(topic, "version")){
+    if (line.ReadChecked(value)){
+      protocol_version = std::min(int(value), XCV_VERSION_2);  // switch protocol version to the minimum version both can do
+    }
+  }
+  return true;
+}
+
+void XVCDevice::LinkTimeout()
+{
+  protocol_version = XCV_VERSION_UNKNOWN;
+  link_timer = 5;
+  nmea_channel_up = false;
+  nmea_enable = false;
+}
+
+bool XVCDevice::EnableNMEA([[maybe_unused]] OperationEnvironment &env)
+{
+  nmea_enable = true;
+  return true;
+}
+
 bool
 XVCDevice::ParseNMEA(const char *String, NMEAInfo &info)
 {
   if (!VerifyNMEAChecksum(String))
     return false;
-
   NMEAInputLine line(String);
   char type[16];
-  line.Read(type, 16);
-
-  if (StringIsEqual(type, "$PXCV"))
-    return PXCV(line, info);
-  else
-    return false;
+  line.Read(type, sizeof(type));
+  if (StringIsEqual(type, "$PXCV")){                // cyclic data from device useful for channel supervision
+    nmea_channel_up = true;
+    if(protocol_version != XCV_VERSION_UNKNOWN){  // only parse NMEA once protocol version is set
+      return PXCV(line, info);
+    }
+    return true;
+  }
+  else if(StringIsEqual(type, "!xcv")){
+    return XCV(line, info);
+  }
+  return false;
 }
 
 // For documentation refer to chapter 10.1.3 Device Driver/XCVario in mulilingual handbook: https://xcvario.de/handbuch
+// Scedule sending of new protocol version 2; peer shall answer with the same message with its version
+// Upon reception of XCV's version, the protocol will be switched, atfer timeout we assume legacy version 1
+void
+XVCDevice::OnSysTicker()
+{
+  // one all us up, repeat sending protocol version every second until timeout ( 5 seconds )
+  if(link_timer && nmea_enable && protocol_version == XCV_VERSION_UNKNOWN && nmea_channel_up){
+    link_timer--;
+    if(protocol_version == XCV_VERSION_UNKNOWN){ // sent our version until negotiation is finished
+      NullOperationEnvironment env;
```

Comment:
We can fix this. Am i good using a static PopupOperationEnvironment like devRestart() in Device/device.cpp ?

static PopupOperationEnvironment env;

Reply:
No, that's just a different class for the same bad trick.
`OnSysTicker()` must not block! And as long as you're using `OperationEnvironment`, no matter which implementation, you have blocking code.
`devRestart()` runs in a different context; it creates an `OperationEnvironment` for a different thread.

Reply:
First, i see that any methods to sent NMEA strings need their OperationEnvironment, e.g. port.FullWriteString(checksum, env, timeout), e.g. the ActionInterface class is using e.g. to sent MC the "MessageOperationEnvironment env; devices->PutMacCready(mc, env);", would this be the correct way to address this? 

And second, the OnSysTicker(), should not be used as risk of blocking with any Operations env, and see no other useful method from base class, so to implement a timer to sent this message, am i going in the right direction with the PeriodicTimer class, to and scedule my own OnTimer() callback e.g. like below ? 
PeriodicTimer timer{[this]{ OnTimer(); }};

Reply:
That's just as wrong - timers must not block as well. Blocking would block the entire thread and all other events registered there, and may freeze XCSoar.
Rule of thumb: if you don't already have an `OperationEnvironment`, you must not block.
All functions which take an `OperationEnvironment` can block for an arbitrary amount of time.

Reply:
Okay i see, great. Taking all this lessons learnt, i could imagine to initiate the sending of protocol version in the context of EnableNMEA(OperationEnvironment &env) ? There is an OperationalEnvironment,  i am allowed to block maybe as long until the first NMEA message from device comes in (?), and until timeout i may resent (even in a loop) the protocol version to the device ?

Reply:
You can use `Port::Write()` but it may be a partial write and you are not allowed to retry if the write was really only partial, as that will result in unbounded wait times.
You may block in `EnableNMEA()`, but I would not advise to wait until the device just happens to reply with a version number; that would not freeze XCSoar, but it would delay each connection attempt to the device.

Reply:
And: obtaining the version number has nothing to do with enabling NMEA ...

Reply:
Guess there is no need to wait in the exceptional condition of a partial write. Like an UDP message on TCP, the protocol version polling can be considered as a fire and forget, no blocking and further supervision needed in lower layer, as supervision we do in the application layer by detecting the peers answer. I think it is correct to user just Write() here. 
Look into that asap, and will repeat all test on my both targets and push this, along with any other fixes for your comments next week.

Reply:
Changed consistently to Port::Write() in protocol version resend code.